txpool/src/containers/dependency.rs

```diff
@@ -0,0 +1,482 @@
+                        state.used_by.insert(tx.id());
+                    }
+                }
+            }
+        }
+
+        // insert all coins/contracts that we got from db;
+        self.coins.extend(db_coins.into_iter());
+        self.contracts.extend(db_contracts.into_iter());
+
+        // iterate over all outputs and insert them, marking them as available.
+        for (index, output) in tx.outputs().iter().enumerate() {
+            let is_contract = match output {
+                Output::Coin { .. }
+                | Output::Withdrawal { .. }
+                | Output::Change { .. }
+                | Output::Variable { .. } => None,
+                Output::Contract { input_index, .. } => {
+                    if let Input::Contract { contract_id, .. } = tx.inputs()[*input_index as usize]
+                    {
+                        Some(contract_id)
+                    } else {
+                        panic!("InputIndex for contract should be always correct");
+                    }
+                }
+                Output::ContractCreated { contract_id } => Some(*contract_id),
+            };
+            let utxo_id = UtxoId::new(tx.id(), index as u8);
+            if let Some(contract_id) = is_contract {
+                // insert contract
+                self.contracts.insert(
+                    contract_id,
+                    ContractState {
+                        depth: max_depth,
+                        used_by: HashSet::new(),
+                        origin: Some(utxo_id),
+                        gas_price: tx.gas_price(),
+                    },
+                );
+            } else {
+                // insert output coin inside by_coin
+                self.coins.insert(
+                    utxo_id,
+                    CoinState {
+                        is_spend_by: None,
+                        depth: max_depth,
+                    },
+                );
+            }
+        }
+
+        Ok(removed_tx)
+    }
+
+    pub fn recursively_remove_all_dependencies<'a>(
+        &'a mut self,
+        txs: &'a HashMap<TxId, ArcTx>,
+        tx: ArcTx,
+    ) -> Vec<ArcTx> {
+        let mut removed_tx = vec![tx.clone()];
+
+        // for all outputs recursivly call removal
+        for (index, output) in tx.outputs().iter().enumerate() {
+            // get contract_id or none if it is coin
+            let is_contract = match output {
+                Output::Coin { .. }
+                | Output::Withdrawal { .. }
+                | Output::Change { .. }
+                | Output::Variable { .. } => None,
+                Output::Contract { input_index, .. } => {
+                    if let Input::Contract { contract_id, .. } = tx.inputs()[*input_index as usize]
+                    {
+                        Some(contract_id)
+                    } else {
+                        panic!("InputIndex for contract should be always correct");
```

Comment:
This should be an error and not a panic

Reply:
"InputIndex for contract should be always correct" this is not expected behaviour.